geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/DataStoreConnector.scala

```diff
@@ -0,0 +1,46 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.spark
+
+import org.geotools.data.{DataStore, DataStoreFinder}
+
+/**
+  * Caches accessing of DataStores.
+  * @param writeDataStoreParams
+  */
+case class DataStoreConnector(writeDataStoreParams: Map[String, String]) {
+  @transient lazy val store: DataStore = DataStoreConnector.getOrFindDataStore(this.writeDataStoreParams)
+}
+
+object DataStoreConnector {
+  import org.locationtech.geomesa.spark.CaseInsensitiveMapFix._
+
+  private val writeLock = new Object
+  @volatile private var connectionMap: Map[Map[String, String], DataStore] = Map()
+
+  def getOrFindDataStore(writeDataStoreParams: Map[String, String]): DataStore = {
+    var map = connectionMap
+    map.get(writeDataStoreParams) match {
```

Comment:
We typically use Caffeine loading caches for things like this.  E.g.:

https://github.com/locationtech/geomesa/blob/master/geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaFeatureSource.scala#L174-L177

I'd suggest that since it avoids implementing the careful synchronized code here.  Also, if we find it useful, it'd be possible to add cache configuration to the DataStoreConnector object.  (I.e., connection could timeout after 10 minutes, etc.)